### PR TITLE
replaced generic to box dyn

### LINKED ISSUE #9441: replace network generic with box dyn
### Describe the feature

the network type is generic because it needs:

https://github.com/paradigmxyz/reth/blob/4e5162cb3f042386244b9268bacb8b6cbe3b5293/crates/net/network/src/swarm.rs#L59-L59

https://github.com/paradigmxyz/reth/blob/4e5162cb3f042386244b9268bacb8b6cbe3b5293/crates/net/network/src/state.rs#L57-L61

now that we have a noop impl available out of the box

https://github.com/paradigmxyz/reth/blob/4e5162cb3f042386244b9268bacb8b6cbe3b5293/crates/net/network/src/config.rs#L454-L460

we can remove this generic entirely, this is only used once per new session, and the value isn't even that useful, so we should simply replace this with `Box<dyn BlockNumReader>` which would make things a bit easier.

## TODO
* replace generic with box dyn

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by loocapro**
Closes: https://github.com/paradigmxyz/reth/issues/9441

### CODE REVIEW

File: crates/net/network/src/manager.rs [hunk 1]
```diff
@@ -112,6 +112,7 @@ pub struct NetworkManager<C> {
     metrics: NetworkMetrics,
     /// Disconnect metrics for the Network
     disconnect_metrics: DisconnectMetrics,
+    _phantom: std::marker::PhantomData<C>,
```
Comment (mattsse) [inline crates/net/network/src/manager.rs:115]: we should be able to get rid of this as well, if not possible due to some downstream requirements, I'm okay with keeping it in this pr

File: crates/net/network/src/state.rs [hunk 2]
```diff
@@ -36,6 +37,30 @@ use tracing::{debug, trace};
 /// Cache limit of blocks to keep track of for a single peer.
 const PEER_BLOCK_CACHE_LIMIT: u32 = 512;
 
+/// Wrapper type for the [`BlockNumReader`] trait.
+pub struct BlockNumReader(Box<dyn reth_storage_api::BlockNumReader>);
```
Comment (mattsse) [inline crates/net/network/src/state.rs:41]:
```suggestion
pub(crate) struct BlockNumReader(Box<dyn reth_storage_api::BlockNumReader>);
```

File: crates/net/network/src/state.rs [hunk 3]
```diff
@@ -36,6 +37,30 @@ use tracing::{debug, trace};
...
+impl BlockNumReader {
+    /// Create a new instance with the given reader.
+    pub fn new(reader: impl reth_storage_api::BlockNumReader + 'static) -> Self {
+        Self(Box::new(reader))
+    }
+}
+
+impl fmt::Debug for BlockNumReader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "<dyn BlockNumReader>")
```
Comment (mattsse) [inline crates/net/network/src/state.rs:52]: should use fmt syntax for formatting structs: ` f.debug_struct(`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/manager.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/network/src/manager.rs:115]: can we not move the generic to the function itself:

```
 pub async fn new<C>(config: NetworkConfig<C>)

```

**Review by mattsse — CHANGES_REQUESTED** [review]
I think we can get of the generic on the struct entirely and only enforce it on the functions that operate on Networkconfig

**Review by mattsse — APPROVED** [review]
nice!

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**